### DISCUSSION TITLE
startNotifications() in background?

### DISCUSSION THREAD

**Discussion by FrancescoPaiola** [capacitor-community/bluetooth-le#514] [Q&A]
Does startNotifications() work while app is in background?
Is there any way to keep listening BLE signals while app is in background?

Thanks in advance.

**Comment by pwespi** [comment]
I have never tested `startNotification` while the app is in the background, but receiving notification in the background works.

Note that on iOS you need to add `bluetooth-central` to the `UIBackgroundModes` in `Info.plist`.

**Comment by crebuh** [comment]
For me this does not work. The listener stops working after about 5 minutes. I already tried to add a foreground service on Android without success. On iOS I'm experiencing the same issue. Any ideas how to keep the listener alive for a little bit longer (1 or 2 hours?)

**Reply by pwespi** [reply]
Hello. I should have been more precise in my answer above. What works is receiving notifications when the app is in the background *but still running* (i.e. another app is currently in the foreground, but the app process is still running).

However if the user closes the app or the OS suspends it, then that's a different story. And I assume this is what happens after about 5 minutes in your case. I have never had this requirement to be honest, therefore I'm not sure if and how this is possible.

**Reply by mutasimissa** [reply]
I'm not sure if this is helpful, but newer iOS and even Adnroid versions have different policies regarding suspending active background processes, I stumbled with that with a case of doing local push notifications, but I don't think it's limited only to push notifications...

**Reply by peitschie** [reply]
There are a few execution limits you'll hit on Android

* At 5mins, there's a WebView optimization that kicks in after the view is hidden that stops the main event loop. To execute beyond this, us this plugin: https://www.npmjs.com/package/@anuradev/capacitor-background-mode and call `disableWebViewOptimizations`
* Sometime around 40mins, Android's own power throttling kicks in (battery optimization) and will put the app into doze mode. To continue executing through this, you need to get the user to disable batter optimizations (a helper is available on the above plugin)
* Neither of these will help if the user closes your app in the background. For that, you need a headless WebView, and a moderate amount of fiddling, also you need to leverage Android's WorkManager.

I've tackled this on a few apps now, with mixed success. See these issues for a wider discussion on the bits... but the above summary is as complete as I understand
* https://github.com/ionic-team/capacitor/issues/6234
* https://github.com/ionic-team/capacitor/issues/3032

On iOS, the UIBackgroundModes is sufficient, just be aware that your app only has about 5-6 seconds to respond to a Bluetooth event when it's woken back up before the app will be terminated again. iOS state restoration might be needed to ensure this functions correctly... I'm not sure if this plugin has the feature yet? (I authored a similar [feature ](https://github.com/don/cordova-plugin-ble-central/pull/888) on another BLE plugin I'm a maintainer of, just as disclaimer! 😅 )
